### DISCUSSION TITLE
PLI Copybooks and reflection of changes

### DISCUSSION THREAD

**Discussion by kraabrsg** [IBM/zopeneditor-about#210] [Q&A] [answered]
Hi, i am on the newest Z open editor (1.4.1) and i wonder when changes in a host dataset (referenced via %include) should be reflected in VSCODIUM ???
1.) Edit a dataset on the host
2.) hover over the include -> no changes
3.) close and open the source file with the include, hover over the include -> no changes
4.) restart VSCODIUM, hover over the include -> changes are present

While it ist not a big deal, because our builds are using the "corrected/ new" versions of copybooks(on the host), the shown seems to be updated only on restart.

Local files:
C:\Users\myuser\.vscode-oss\extensions\ibm.zopeneditor-1.4.1\resources\temp\_D_\profilename\HOST.PO.DS

(Maybe related to: #108 )

Thanks!

**Comment by FALLAI-Denis** [comment]
Hi,

I think this problem concerns Zowe Exporer more than Z Open Editor.
Z Open Editor uses Zowe Explorer to download the COPYs and INCLUDEs, and Zowe Explorer provides temporary local storage.
The temporary folder is cleaned when Zowe Explorer is unloaded.
See Zowe Explorer settings.

![image](https://user-images.githubusercontent.com/63861965/148549345-e4ed68a0-7f35-40ae-a0de-2c096ae509ab.png)


Perhaps Zowe Explorer should have a "time to live" for locally temporary downloaded files.

**Comment by phaumer** [comment]
@kraabrsg would you be able to attach a [DEBUG level log file](https://ibm.github.io/zopeneditor-about/Docs/locating_local_client_logs.html) for this scenario? The file should be downloaded after you typed the name after the %include. If you type slowly you would see in the log that tries to get it for the partial name, but it should try again when the name is complete and you continue typing, e.g. adding a semi colon or return.

We do have one open issue for very slow connections right now that we need to fix: https://github.com/IBM/zopeneditor-about/issues/197

**Reply by kraabrsg** [reply]
Hi @phaumer i attached a DEBUG file.
1.) i typed slowly
2.) closed and opened the test pli file 2 times.
3.) tested with the files reported in #211 

[zopeneditor-20220111-062901.log](https://github.com/IBM/zopeneditor-about/files/7852133/zopeneditor-20220111-062901.log)

**Reply by phaumer** [reply]
@kraabrsg thanks. When looking at the log file I can see that the editor starts and that it finds include file in `KTEST2` in `myuser.DEV.INC(KTEST2)` right away. It then tries to locate `KTEST` without the 2 in `myuser.DEV.INC(KTEST)`, `???.SCM.IDE.COPYBOOK(KTEST)`, `???.BBS.T555.COPYBOOK(KTEST)`, `???.BBS.B555.COPYBOOK(KTEST)`, and `???.BBS.R555.COPYBOOK(KTEST)`. All of these return a 404/not found error. Is that incorrect? Should one of these been a Found? Or was the problem that you typed `KTEST2` again and it cut off the 2 always just searching for KTEST? One request was also for `KTES`, which I assume was the slow typing. 

When I try this on my end slowly typing an include file name from MVS using z/OSMF I do get 404 for the incomplete name, then when the name is complete it finds it and I see briefly a red squiggle as well as the underline link. As soon as I type the next character, which is the `;` then then red squiggle goes away.

**Comment by kraabrsg** [comment]
Hi @phaumer you nailed it, only KTEST2 should be found. It is working like expected. My initial question was about reflecting updates. Because the host file is downloaded locally, it does not reflect changes on the host file, only on restart of vscodium.  

Please remove the not anonymized PO names (starting with RZ...)in your comments, i updated the log file, i think personally it does not contain sensible data, but i am advised to fully "scrabble" log files and information.

**Comment by phaumer** [comment, marked as answer]
@kraabrsg oh, I misunderstood. Yes, that is an open issue: when the files change remotely we do not recognize it, nor is there a manual trigger to re-download as we have in Zowe Explorer.  We need to add that. Also people ask about the other way round: if they open these downloaded files by clicking on the hyperlink in the program and make changes that those get written back to MVS. Also something we still need to add.

**Comment by kraabrsg** [comment]
Hi @phaumer , as for pli, jcl and copybook files itself we use an extension called "trigger task on save" there we narrow down file extension, when saved in vscode a task is started that does the upload to MVS(powershell)... It is not really fast now, but the daemon mode in zowe V2 will speed that up dramatically. 

When implementing such a feature to write back to MVS please add some kind of filter setting. 
In our case, we are allowed to edit only copybooks that our on our private libs([user].xxx.yyy). The other PO Datasets in the propertygrous are write protected  for the users, so this would end up in somekind of zosmf errors. 
E,g. -> a setting like "Allowed_Origin_To_Be_Updated_Which_Resides_On = xxx.yy.*"

 Further please completely scramble the PO Dataset in your post.

Feedback is highly appreciated and valuable, thank you!

**Reply by FALLAI-Denis** [reply]
We are using Git, and all copybooks must be updated thru Git process (and Jenkins pipeline). 
copybooks are duplicated on z/os for compilation purpose and are searched by VS Code for out of scope of current workspace / Git repository (interface with other application). 
we do’nt want to autorize direct update of COPYBOOK on z/os, (same for cobol source also duplicated on z/os for other reasons). 

So direct update must be forbidden by Z Open Editor settings : full or by dsname / path filter. 

I suggest to add a dedicated property on each property-group entry for that (for « mvs » type only?):
allowupdate: true/false

**Reply by phaumer** [reply]
Good points and ideas. Thanks.
